src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java

```diff
@@ -358,6 +371,123 @@ public void execute() throws MojoExecutionException, MojoFailureException {
         registerSourceRoots();
     }
 
+    private void doExec(final String bootClassName, final SecurityManager originalSecurityManager) throws Throwable {
+        Class<?> bootClass = Thread.currentThread().getContextClassLoader().loadClass(bootClassName);
+        MethodHandles.Lookup lookup = MethodHandles.lookup();
+        try {
+            doMain(
+                    originalSecurityManager,
+                    lookup.findStatic(bootClass, "main", MethodType.methodType(void.class, String[].class)));
+        } catch (final NoSuchMethodException nsme) {
+            if (Runnable.class.isAssignableFrom(bootClass)) {
+                doRun(bootClass);
+            } else {
+                throw nsme;
+            }
+        }
+    }
+
+    private void doMain(final SecurityManager originalSecurityManager, final MethodHandle mainHandle) throws Throwable {
+        if (blockSystemExit && originalSecurityManager == null) {
+            System.setSecurityManager(new SystemExitManager(originalSecurityManager));
+        }
+        mainHandle.invoke(arguments);
+    }
+
+    private void doRun(final Class<?> bootClass)
+            throws InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchMethodException {
+        final Class<? extends Runnable> runnableClass = bootClass.asSubclass(Runnable.class);
+        final Constructor<? extends Runnable> constructor = Stream.of(runnableClass.getDeclaredConstructors())
+                .map(i -> (Constructor<? extends Runnable>) i)
+                .filter(i -> Modifier.isPublic(i.getModifiers()))
+                .max(Comparator.<Constructor<? extends Runnable>, Integer>comparing(Constructor::getParameterCount))
+                .orElseThrow(() -> new IllegalArgumentException("No public constructor found for " + bootClass));
+        if (getLog().isDebugEnabled()) {
+            getLog().debug("Using constructor " + constructor);
+        }
+
+        Runnable runnable;
+        try { // todo: enhance that but since injection API is being defined at mvn4 level it is
+            // good enough
+            final Object[] args = Stream.of(constructor.getParameters())
+                    .map(param -> {
+                        try {
+                            return lookupParam(param);
+                        } catch (final ComponentLookupException e) {
+                            getLog().error(e.getMessage(), e);
+                            throw new IllegalStateException(e);
+                        }
+                    })
+                    .toArray(Object[]::new);
+            constructor.setAccessible(true);
```

Comment:
We filter by `public` modifier - so is it needed?

Reply:
Yes due to cross classloader behavior we can have, visibility is not a sufficient guarantee